res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,772 @@
+                MixtrackProFX.shifted = false;
+                MixtrackProFX.deck[0].unshift();
+                MixtrackProFX.deck[1].unshift();
+                MixtrackProFX.browse.unshift();
+                MixtrackProFX.effect.unshift();
+            }
+
+            // for displaying pads lights when shifted
+            MixtrackProFX.deck[0].pads.reconnectComponents();
+            MixtrackProFX.deck[1].pads.reconnectComponents();
+        }
+    });
+
+    this.loop = new components.Button({
+        key: "loop_enabled",
+        midi: [0x94 + channel, 0x40],
+        off: 0x01,
+        input: function(channel, control, value, status, group) {
+            if (engine.getValue(group, "loop_enabled") === 0) {
+                script.triggerControl(group, "beatloop_activate");
+            } else {
+                script.triggerControl(group, "beatlooproll_activate");
+            }
+        }
+    });
+
+    this.reloop = new components.Button({
+        inKey: "reloop_toggle" // or loop_in_goto to not enable loop
+    });
+
+    this.loopHalf = new components.Button({
+        key: "loop_halve",
+        midi: [0x94 + channel, 0x34],
+        off: 0x01
+    });
+
+    this.loopDouble = new components.Button({
+        key: "loop_double",
+        midi: [0x94 + channel, 0x35],
+        off: 0x01
+    });
+
+    this.loopIn = new components.Button({
+        inKey: "loop_in"
+    });
+
+    this.loopOut = new components.Button({
+        inKey: "loop_out"
+    });
+
+    // powerWindow workaround
+    this.bleep = new components.Button({
+        input: function(channel, control, value, status, group) {
+            if (value === 0x7F) {
+                value = 1;
+            }
+            engine.setValue(group, "reverseroll", value);
+        }
+    });
+
+    // powerWindow workaround
+    this.pitchBendUp = new components.Button({
+        input: function(channel, control, value, status, group) {
+            if (value === 0x7F) {
+                value = 1;
+            }
+            engine.setValue(group, "rate_temp_up", value);
+        }
+    });
+
+    // powerWindow workaround
+    this.pitchBendDown = new components.Button({
+        input: function(channel, control, value, status, group) {
+            if (value === 0x7F) {
+                value = 1;
+            }
+            engine.setValue(group, "rate_temp_down", value);
+        }
+    });
+
+    this.keylock = new components.Button({
+        type: components.Button.prototype.types.toggle,
+        inKey: "keylock"
+    });
+
+    this.pitchRange = new components.Button({
+        currentRangeIdx: 0,
+        input: function(channel, control, value, status, group) {
+            this.currentRangeIdx = (this.currentRangeIdx + 1) % MixtrackProFX.pitchRanges.length;
+            engine.setValue(group, "rateRange", MixtrackProFX.pitchRanges[this.currentRangeIdx]);
+        }
+    });
+
+    this.prevEffect = new components.Button({
```

Comment:
Because each deck has its own prev/next FX buttons. Should I put those buttons it in a separate ComponentContainer? I suppose it would be more logical.

Reply:
MixtrackProFX.EffectUnit is probably the best place.

Reply:
Yes, ideally you would group controls in code similar to how they're grouped on the hardware. You can also nest ComponentContainers if that makes sense.